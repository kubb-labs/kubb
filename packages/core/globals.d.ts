/**
* `tsconfig.json`
* @example
"compilerOptions": {
___ "types": ["@kubb/core/globals"]
}
* @example implementation
type SwaggerPlugin = Kubb.Plugins["@kubb/swagger"]
const swaggerPlugin: SwaggerPlugin={
___ validate: true
}
*/
declare namespace Kubb {
  namespace Helpers {
    type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never
    type LastOf<T> = UnionToIntersection<T extends any ? () => T : never> extends () => infer R ? R : never

    // TS4.0+
    type Push<T extends any[], V> = [...T, V]

    // TS4.1+
    type TuplifyUnion<T, L = LastOf<T>, N = [T] extends [never] ? true : false> = true extends N ? [] : Push<TuplifyUnion<Exclude<T, L>>, L>

    type ObjValueTuple<T, KS extends any[] = TuplifyUnion<keyof T>, R extends any[] = []> = KS extends [infer K, ...infer KT]
      ? ObjValueTuple<T, KT, [...R, [name: K & keyof T, options: T[K & keyof T]]]>
      : R

    type TupleToUnion<T> = T extends Array<infer ITEMS> ? ITEMS : never
  }
  type Plugins = {
    ['@kubb/plugin-oas']: import('@kubb/plugin-Oas').PluginOas
    ['@kubb/plugin-redoc']: import('@kubb/plugin-redoc').PluginRedoc
    ['@kubb/swagger-client']: import('@kubb/swagger-client').PluginClient
    ['@kubb/swagger-faker']: import('@kubb/swagger-faker').PluginFaker
    ['@kubb/swagger-swr']: import('@kubb/swagger-swr').PluginSwr
    ['@kubb/swagger-tanstack-query']: import('@kubb/swagger-tanstack-query').PluginTanstackQuery
    ['@kubb/swagger-ts']: import('@kubb/swagger-ts').PluginTs
    ['@kubb/swagger-zod']: import('@kubb/swagger-zod').PluginZod
    ['@kubb/swagger-zodios']: import('@kubb/swagger-zodios').PluginZodios
    ['@kubb/swagger-msw']: import('@kubb/swagger-msw').PluginMsw
  }
  type OptionsPlugins = { [K in keyof Plugins]: Plugins[K]['options'] }

  type OptionsOfPlugin<K extends keyof Plugins> = Plugins[K]['options']

  type PluginUnion = Helpers.TupleToUnion<Helpers.ObjValueTuple<OptionsPlugins>>

  type Plugin = keyof Plugins
}
