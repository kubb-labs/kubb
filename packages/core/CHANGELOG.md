# @kubb/core

## 2.11.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.11.1
  - @kubb/types@2.11.1

## 2.11.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.11.0
  - @kubb/types@2.11.0

## 2.10.0

### Minor Changes

- [#878](https://github.com/kubb-project/kubb/pull/878) [`a80cc2d`](https://github.com/kubb-project/kubb/commit/a80cc2d2a4f109ff1e814707e5dd104bd730fb64) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - support for useQuery for post/put/delete/patch methods

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.10.0
  - @kubb/types@2.10.0

## 2.9.1

### Patch Changes

- [`6109ffa`](https://github.com/kubb-project/kubb/commit/6109ffa77aa9e6d629eff06850d2fe4bcd62088c) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - update packages

- Updated dependencies []:
  - @kubb/parser@2.9.1
  - @kubb/types@2.9.1

## 2.9.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.9.0
  - @kubb/types@2.9.0

## 2.8.2

### Patch Changes

- Updated dependencies [[`9759907`](https://github.com/kubb-project/kubb/commit/9759907ac07abd69021712666331f34fc8fa33f0)]:
  - @kubb/parser@2.8.2
  - @kubb/types@2.8.2

## 2.8.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.8.1
  - @kubb/types@2.8.1

## 2.8.0

### Patch Changes

- [#852](https://github.com/kubb-project/kubb/pull/852) [`752f9a0`](https://github.com/kubb-project/kubb/commit/752f9a02f642f2c5e948b96622fdc73c33d571b8) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - useOperation and useSchema with a component Oas and Oas.Operation

- Updated dependencies []:
  - @kubb/parser@2.8.0
  - @kubb/types@2.8.0

## 2.7.2

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.7.2
  - @kubb/types@2.7.2

## 2.7.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.7.1
  - @kubb/types@2.7.1

## 2.7.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.7.0
  - @kubb/types@2.7.0

## 2.6.7

### Patch Changes

- [`93940b3`](https://github.com/kubb-project/kubb/commit/93940b34dbe0c0ad5e81db3c0db5c40dfeed380f) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - expose all core types

- Updated dependencies []:
  - @kubb/parser@2.6.7
  - @kubb/types@2.6.7

## 2.6.6

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.6
  - @kubb/types@2.6.6

## 2.6.5

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.5
  - @kubb/types@2.6.5

## 2.6.4

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.4
  - @kubb/types@2.6.4

## 2.6.3

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.3
  - @kubb/types@2.6.3

## 2.6.2

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.2
  - @kubb/types@2.6.2

## 2.6.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.1
  - @kubb/types@2.6.1

## 2.6.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.6.0
  - @kubb/types@2.6.0

## 2.5.3

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.5.3
  - @kubb/types@2.5.3

## 2.5.2

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.5.2
  - @kubb/types@2.5.2

## 2.5.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.5.1
  - @kubb/types@2.5.1

## 2.5.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.5.0
  - @kubb/types@2.5.0

## 2.4.1

### Patch Changes

- [#796](https://github.com/kubb-project/kubb/pull/796) [`bf1e521`](https://github.com/kubb-project/kubb/commit/bf1e52182b636cef8c2cd54bcb123baf47d2d624) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - A schema name with `pet.Pets` will be moved inside of the folder `src/gen/pet/pets/ts` instead of `src/gen/petPets.ts`

- Updated dependencies []:
  - @kubb/parser@2.4.1
  - @kubb/types@2.4.1

## 2.4.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.4.0
  - @kubb/types@2.4.0

## 2.3.0

### Patch Changes

- Updated dependencies [[`ec9c07d`](https://github.com/kubb-project/kubb/commit/ec9c07d90eb3472f5d0030a1cbb746e0055b8ab8)]:
  - @kubb/parser@2.3.0
  - @kubb/types@2.3.0

## 2.2.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.2.1
  - @kubb/types@2.2.1

## 2.2.0

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.2.0
  - @kubb/types@2.2.0

## 2.1.6

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.1.6
  - @kubb/types@2.1.6

## 2.1.5

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.1.5
  - @kubb/types@2.1.5

## 2.1.4

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.1.4
  - @kubb/types@2.1.4

## 2.1.3

### Patch Changes

- [`9307bda`](https://github.com/kubb-project/kubb/commit/9307bda2c2dc08503809eec7d048bba4e6388121) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - use of Factory['response'] instead of data to support `dataReturnType`

- Updated dependencies []:
  - @kubb/parser@2.1.3
  - @kubb/types@2.1.3

## 2.1.2

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.1.2
  - @kubb/types@2.1.2

## 2.1.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.1.1
  - @kubb/types@2.1.1

## 2.1.0

### Minor Changes

- [#713](https://github.com/kubb-project/kubb/pull/713) [`c22433e`](https://github.com/kubb-project/kubb/commit/c22433ec2c1b04527d72a6de44e524e580e2d876) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - `barrelNamed` for `exportType` to use `export { nameX, nameY } from '.'`

### Patch Changes

- Updated dependencies [[`c22433e`](https://github.com/kubb-project/kubb/commit/c22433ec2c1b04527d72a6de44e524e580e2d876)]:
  - @kubb/parser@2.1.0
  - @kubb/types@2.1.0

## 2.0.6

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.0.6
  - @kubb/types@2.0.6

## 2.0.5

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.0.5
  - @kubb/types@2.0.5

## 2.0.4

## 2.0.3

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.0.3
  - @kubb/types@2.0.3

## 2.0.2

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.0.2
  - @kubb/types@2.0.2

## 2.0.1

### Patch Changes

- Updated dependencies []:
  - @kubb/parser@2.0.1
  - @kubb/types@2.0.1

## 2.0.0

### Major Changes

- [#686](https://github.com/kubb-project/kubb/pull/686) [`0c894ca`](https://github.com/kubb-project/kubb/commit/0c894ca935045272a3427ed5646a83184646e354) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - swagger-ts with output object

- [#676](https://github.com/kubb-project/kubb/pull/676) [`d729470`](https://github.com/kubb-project/kubb/commit/d729470b74121eef6776649654921ce61b35da51) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - transformers subpackage

- [#678](https://github.com/kubb-project/kubb/pull/678) [`48b7ff2`](https://github.com/kubb-project/kubb/commit/48b7ff246a3459bb7a9be6d430407c2538d3b2eb) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - swagger `Infer<typeof oas>` type

### Minor Changes

- [`210d58f`](https://github.com/kubb-project/kubb/commit/210d58fd1fcc1e8d84f38fdfabbb59630a7394b5) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - `exportType` to disable the creation of barrel files

### Patch Changes

- [#707](https://github.com/kubb-project/kubb/pull/707) [`955f8ed`](https://github.com/kubb-project/kubb/commit/955f8edc26ca303f3432ed875a97e249c88df89b) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - ignore ext when using output.path and output.extname is not set

- [#676](https://github.com/kubb-project/kubb/pull/676) [`d729470`](https://github.com/kubb-project/kubb/commit/d729470b74121eef6776649654921ce61b35da51) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - add 'root' to KubbFile.Import

- [#689](https://github.com/kubb-project/kubb/pull/689) [`8044907`](https://github.com/kubb-project/kubb/commit/8044907f560f1e9a6120df259568b9213a4f1e4a) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - path.extName to set `.ts` or `.js` to the barrel files

- [#674](https://github.com/kubb-project/kubb/pull/674) [`6348057`](https://github.com/kubb-project/kubb/commit/634805723409381eace8e68fd5f2eab6f737dd7a) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - pluginSorter to sort based on pre and post

- [#707](https://github.com/kubb-project/kubb/pull/707) [`955f8ed`](https://github.com/kubb-project/kubb/commit/955f8edc26ca303f3432ed875a97e249c88df89b) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - use of combineFiles

- [#707](https://github.com/kubb-project/kubb/pull/707) [`955f8ed`](https://github.com/kubb-project/kubb/commit/955f8edc26ca303f3432ed875a97e249c88df89b) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - use of `p-queue` instead of our own queue

- [`e17bc7c`](https://github.com/kubb-project/kubb/commit/e17bc7ccfb91aeab52488e847356890464aa6166) Thanks [@stijnvanhulle](https://github.com/stijnvanhulle)! - tsup build cleanup with `splitting`

- Updated dependencies [[`0c894ca`](https://github.com/kubb-project/kubb/commit/0c894ca935045272a3427ed5646a83184646e354), [`48b7ff2`](https://github.com/kubb-project/kubb/commit/48b7ff246a3459bb7a9be6d430407c2538d3b2eb), [`d729470`](https://github.com/kubb-project/kubb/commit/d729470b74121eef6776649654921ce61b35da51), [`e17bc7c`](https://github.com/kubb-project/kubb/commit/e17bc7ccfb91aeab52488e847356890464aa6166)]:
  - @kubb/types@2.0.0
  - @kubb/parser@2.0.0
