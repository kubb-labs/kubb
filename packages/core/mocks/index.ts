import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
