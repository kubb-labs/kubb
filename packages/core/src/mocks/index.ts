import path from 'node:path'
import { getSource } from '../FileManager'
import type { File, ResolvedFile } from '../fs/types.ts'
import type { Logger } from '../logger'
import type { PluginManager } from '../PluginManager.ts'
import { camelCase, pascalCase } from '../transformers/casing.ts'
import type { Plugin } from '../types.ts'
import { createFile } from '../utils'

export const mockedLogger = {
  emit(_type, _message) {},
  on(_type, _message) {},
  consola: {},
} as Logger

export const createMockedPluginManager = (name?: string) =>
  ({
    resolveName: (result) => {
      if (result.type === 'file') {
        return camelCase(name || result.name)
      }

      if (result.type === 'type') {
        return pascalCase(result.name)
      }

      if (result.type === 'function') {
        return camelCase(result.name)
      }

      return camelCase(result.name)
    },
    config: {
      output: {
        path: './path',
      },
    },
    resolvePath: ({ baseName }) => baseName,
    logger: {
      emit(message) {
        console.log(message)
      },
      on(_eventName, _args) {},
      logLevel: 3,
    },
    getPluginByKey: (_pluginKey: Plugin['key']) => {
      return undefined
    },
    getFile: ({ name, extname, pluginKey }) => {
      const baseName = `${name}${extname}`

      return {
        path: baseName,
        baseName,
        meta: {
          pluginKey,
        },
      }
    },
  }) as PluginManager

export const mockedPluginManager = createMockedPluginManager('')

export async function matchFiles(files: Array<ResolvedFile | File> | undefined, pre?: string) {
  if (!files) {
    return undefined
  }

  for await (const file of files) {
    const source = await getSource(createFile(file), { logger: mockedLogger })
    await expect(source).toMatchFileSnapshot(path.join(...['__snapshots__', pre, file.path].filter(Boolean)))
  }
}
