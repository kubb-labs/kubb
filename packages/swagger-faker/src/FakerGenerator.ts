import { SchemaGenerator } from '@kubb/swagger'
import { pluginKey as swaggerTypeScriptPluginKey } from '@kubb/swagger-ts'

import { fakerParser } from './fakerParser.tsx'
import { pluginKey } from './plugin.ts'

import type { SchemaGeneratorBuildOptions, SchemaGeneratorOptions } from '@kubb/swagger'

type Options = SchemaGeneratorOptions & {
  seed?: number | number[]
}

export class FakerGenerator extends SchemaGenerator<Options> {
  build({
    schema,
    name: baseName,
    operationName,
    operation,
    description,
  }: SchemaGeneratorBuildOptions): string[] {
    const texts: string[] = []
    const input = this.getTypeFromSchema(schema, baseName)

    const name = this.context.pluginManager.resolveName({ name: baseName, pluginKey, type: 'function' })
    const typeName = this.context.pluginManager.resolveName({ name: baseName, pluginKey: swaggerTypeScriptPluginKey, type: 'type' })

    const output = fakerParser(input, {
      name,
      typeName,
      description,
      seed: this.options.seed,
    })
    // hack to add typescript imports
    if (typeName) {
      const typeFileName = this.context.pluginManager.resolveName({ name: baseName, pluginKey: swaggerTypeScriptPluginKey, type: 'file' })

      const typePath = this.context.pluginManager.resolvePath({
        baseName: operationName || typeFileName,
        pluginKey: swaggerTypeScriptPluginKey,
        options: { tag: operation?.getTags()[0]?.name },
      })

      if (typePath) {
        this.imports.push({
          ref: {
            propertyName: typeName,
            originalName: baseName,
            pluginKey: swaggerTypeScriptPluginKey,
          },
          path: typePath,
          isTypeOnly: true,
        })
      }
    }

    texts.push(output)

    return texts
  }
}
