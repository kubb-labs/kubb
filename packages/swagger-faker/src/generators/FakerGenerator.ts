import { SchemaGenerator } from '@kubb/core'
import { getUniqueName, transformers } from '@kubb/core/utils'
import { isReference } from '@kubb/swagger/utils'
import { pluginKey as swaggerTypeScriptPluginKey } from '@kubb/swagger-ts'

import { fakerKeywords, fakerParser } from '../parsers/index.ts'
import { pluginKey } from '../plugin.ts'

import type { PluginContext } from '@kubb/core'
import type { FileResolver, ImportMeta, OpenAPIV3, Refs } from '@kubb/swagger'
import type ts from 'typescript'
import type { FakerKeyword, FakerMeta } from '../parsers/index.ts'

type Options = {
  fileResolver?: FileResolver
  withJSDocs?: boolean
  resolveName: PluginContext['resolveName']
  dateType: 'string' | 'date'
}
export class FakerGenerator extends SchemaGenerator<Options, OpenAPIV3.SchemaObject, string[]> {
  // Collect the types of all referenced schemas so we can export them later
  refs: Refs = {}

  imports: ImportMeta[] = []

  extraTexts: string[] = []

  aliases: ts.TypeAliasDeclaration[] = []

  // Keep track of already used type aliases
  #usedAliasNames: Record<string, number> = {}

  constructor(options: Options = { withJSDocs: true, dateType: 'string', resolveName: ({ name }) => name }) {
    super(options)

    return this
  }

  build({
    schema,
    baseName,
    description,
    operationName,
  }: {
    schema: OpenAPIV3.SchemaObject
    baseName: string
    description?: string
    operationName?: string
  }): string[] {
    const texts: string[] = []
    const fakerInput = this.#getTypeFromSchema(schema, baseName)
    if (description) {
      texts.push(transformers.JSDoc.createJSDocBlockText({ comments: [`@description ${description}`] }))
    }

    const name = this.options.resolveName({ name: baseName, pluginKey }) || baseName
    const typeName = this.options.resolveName({ name: baseName, pluginKey: swaggerTypeScriptPluginKey })

    const fakerOutput = fakerParser(fakerInput, {
      name,
      typeName,
    })
    // hack to create import with recreating the ImportGenerator
    if (typeName) {
      const ref = {
        propertyName: typeName,
        originalName: baseName,
        pluginKey: swaggerTypeScriptPluginKey,
      }
      this.imports.push({
        ref,
        path: this.options.fileResolver?.(operationName || typeName, ref) || '',
      })
    }

    texts.push(fakerOutput)

    return [...this.extraTexts, ...texts]
  }

  /**
   * Creates a type node from a given schema.
   * Delegates to getBaseTypeFromSchema internally and
   * optionally adds a union with null.
   */
  #getTypeFromSchema(schema: OpenAPIV3.SchemaObject, baseName?: string): FakerMeta[] {
    const validationFunctions = this.#getBaseTypeFromSchema(schema, baseName)
    if (validationFunctions) {
      return validationFunctions
    }

    return []
  }

  /**
   * Recursively creates a type literal with the given props.
   */
  #getTypeFromProperties(baseSchema?: OpenAPIV3.SchemaObject, _baseName?: string): FakerMeta[] {
    const properties = baseSchema?.properties || {}
    const additionalProperties = baseSchema?.additionalProperties

    const objectMembers = Object.keys(properties)
      .map((name) => {
        const validationFunctions: FakerMeta[] = []

        const schema = properties[name] as OpenAPIV3.SchemaObject

        validationFunctions.push(...this.#getTypeFromSchema(schema, name))

        return {
          [name]: validationFunctions,
        }
      })
      .reduce((acc, curr) => ({ ...acc, ...curr }), {})

    const members: FakerMeta[] = []

    if (additionalProperties) {
      const addionalValidationFunctions: FakerMeta[] = additionalProperties === true
        ? [{ keyword: fakerKeywords.any }]
        : this.#getTypeFromSchema(additionalProperties as OpenAPIV3.SchemaObject)

      members.push({ keyword: fakerKeywords.catchall, args: addionalValidationFunctions })
    }

    return [{ keyword: fakerKeywords.object, args: objectMembers }]
  }

  /**
   * Create a type alias for the schema referenced by the given ReferenceObject
   */
  #getRefAlias(obj: OpenAPIV3.ReferenceObject, _baseName?: string): FakerMeta[] {
    const { $ref } = obj
    let ref = this.refs[$ref]

    if (ref) {
      return [{ keyword: fakerKeywords.ref, args: ref.propertyName }]
    }

    const originalName = getUniqueName($ref.replace(/.+\//, ''), this.#usedAliasNames)
    const propertyName = this.options.resolveName({ name: originalName, pluginKey }) || originalName

    ref = this.refs[$ref] = {
      propertyName,
      originalName,
    }

    return [{ keyword: fakerKeywords.ref, args: ref.propertyName }]
  }

  /**
   * This is the very core of the OpenAPI to TS conversion - it takes a
   * schema and returns the appropriate type.
   */
  #getBaseTypeFromSchema(schema: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined, baseName?: string): FakerMeta[] {
    if (!schema) {
      return [{ keyword: fakerKeywords.any }]
    }

    if (isReference(schema)) {
      return this.#getRefAlias(schema, baseName)
    }

    if (schema.oneOf) {
      // union
      const schemaWithoutOneOf = { ...schema, oneOf: undefined }

      const union: FakerMeta = {
        keyword: fakerKeywords.union,
        args: schema.oneOf
          .map((item) => {
            return this.#getBaseTypeFromSchema(item)[0]
          })
          .filter(Boolean)
          .filter((item) => {
            return item && item.keyword !== fakerKeywords.any
          }),
      }
      if (schemaWithoutOneOf.properties && union.args) {
        return [{ ...union, args: [...this.#getBaseTypeFromSchema(schemaWithoutOneOf, baseName), ...union.args] }]
      }

      return [union]
    }

    if (schema.anyOf) {
      // union
      const schemaWithouAnyOf = { ...schema, anyOf: undefined }

      const union: FakerMeta = {
        keyword: fakerKeywords.union,
        args: schema.anyOf
          .map((item) => {
            return this.#getBaseTypeFromSchema(item)[0]
          })
          .filter(Boolean)
          .filter((item) => {
            return item && item.keyword !== fakerKeywords.any
          }),
      }
      if (schemaWithouAnyOf.properties && union.args) {
        return [{ ...union, args: [...this.#getBaseTypeFromSchema(schemaWithouAnyOf, baseName), ...union.args] }]
      }

      return [union]
    }
    if (schema.allOf) {
      // intersection/add
      const schemaWithoutAllOf = { ...schema, allOf: undefined }

      const and: FakerMeta = {
        keyword: fakerKeywords.and,
        args: schema.allOf
          .map((item) => {
            return this.#getBaseTypeFromSchema(item)[0]
          })
          .filter(Boolean)
          .filter((item) => {
            return item && item.keyword !== fakerKeywords.any
          }),
      }

      if (schemaWithoutAllOf.properties && and.args) {
        return [{ ...and, args: [...this.#getBaseTypeFromSchema(schemaWithoutAllOf, baseName), ...and.args] }]
      }

      return [and]
    }

    if (schema.enum) {
      if ('x-enumNames' in schema) {
        return [
          {
            keyword: fakerKeywords.enum,
            args: [`[${[...new Set(schema['x-enumNames'] as string[])].map((value) => `\`${value}\``).join(', ')}]`],
          },
        ]
      }

      if (schema.type === 'number' || schema.type === 'integer') {
        return [
          {
            keyword: fakerKeywords.enum,
            args: [`[${[...new Set(schema.enum)].join(', ')}]`],
          },
        ]
      }

      return [
        {
          keyword: fakerKeywords.enum,
          args: [`[${[...new Set(schema.enum)].map((value: string) => `\`${value}\``).join(', ')}]`],
        },
      ]
    }

    if ('items' in schema) {
      // items -> array
      return [{ keyword: fakerKeywords.array, args: this.#getTypeFromSchema(schema.items as OpenAPIV3.SchemaObject, baseName) }]
    }

    if ('prefixItems' in schema) {
      const prefixItems = schema.prefixItems as OpenAPIV3.SchemaObject[]

      return [
        {
          keyword: fakerKeywords.tuple,
          args: prefixItems
            .map((item) => {
              // no baseType so we can fall back on an union when using enum
              return this.#getBaseTypeFromSchema(item, undefined)?.[0]
            })
            .filter(Boolean),
        },
      ]
    }

    if (schema.properties || schema.additionalProperties) {
      // properties -> literal type
      return this.#getTypeFromProperties(schema, baseName)
    }

    if (schema.type) {
      if (Array.isArray(schema.type)) {
        // OPENAPI v3.1.0: https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0
        const [type] = schema.type as Array<OpenAPIV3.NonArraySchemaObjectType>

        return [
          ...this.#getBaseTypeFromSchema(
            {
              ...schema,
              type,
            },
            baseName,
          ),
          { keyword: fakerKeywords.null },
        ]
      }

      if (schema.type === fakerKeywords.number || schema.type === fakerKeywords.integer) {
        const min = schema.minimum ?? schema.minLength ?? schema.minItems ?? undefined
        const max = schema.maximum ?? schema.maxLength ?? schema.maxItems ?? undefined

        return [{ keyword: fakerKeywords.number, args: { min, max } }]
      }

      if (schema.pattern) {
        return [{ keyword: fakerKeywords.matches, args: `/${schema.pattern}/` }]
      }

      if (this.options.dateType === 'date' && ['date', 'date-time'].some((item) => item === schema.format)) {
        return [{ keyword: fakerKeywords.datetime }]
      }

      if (schema.format === 'email' || baseName === 'email') {
        return [{ keyword: fakerKeywords.email }]
      }

      if (schema.format === 'uri' || schema.format === 'hostname') {
        return [{ keyword: fakerKeywords.url }]
      }
      if (schema.format === 'uuid') {
        return [{ keyword: fakerKeywords.uuid }]
      }

      // based on baseName
      if (baseName === 'firstName') {
        return [{ keyword: fakerKeywords.firstName }]
      }

      if (baseName === 'lastName') {
        return [{ keyword: fakerKeywords.lastName }]
      }

      if (baseName === 'password') {
        return [{ keyword: fakerKeywords.password }]
      }

      if (baseName === 'phone') {
        return [{ keyword: fakerKeywords.phone }]
      }

      // string, boolean, null, number
      if (schema.type in fakerKeywords) {
        return [{ keyword: schema.type as FakerKeyword }]
      }
    }

    if (schema.format === 'binary') {
      // TODO binary
    }

    return [{ keyword: fakerKeywords.any }]
  }
}
