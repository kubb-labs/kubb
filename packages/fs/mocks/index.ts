import { pascalCase } from '../src/transformers/casing.ts'

import { FileManager, type KubbFile } from '../src'
import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extname, pluginKey }) => {
    const baseName = `${name}${extname}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager

export async function matchFiles(files: KubbFile.File[]) {
  for (const file of files) {
    const source = await FileManager.getSource(file)
    expect(source).toMatchSnapshot()
  }
}
