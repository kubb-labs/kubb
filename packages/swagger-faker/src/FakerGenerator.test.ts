import path from 'node:path'

import { mockedPluginManager } from '@kubb/core/mocks'
import { OasManager } from '@kubb/swagger'

import { FakerGenerator } from './FakerGenerator'

import type { Plugin } from '@kubb/core'
import type { SchemaObject } from '@kubb/swagger/oas'
import type { PluginOptions } from './types'

describe('FakeGenerator enums', async () => {
  const schemaPath = path.resolve(__dirname, '../mocks/enums.yaml')
  const oas = await new OasManager().parse(schemaPath)
  const generator = new FakerGenerator({
    dateType: 'string',
    mapper: undefined,
    seed: 1,
    transformers: {},
    unknownType: 'any',
    enumType: 'asConst',
    usedEnumNames: {},
  }, {
    oas,
    pluginManager: mockedPluginManager,
    plugin: {} as Plugin<PluginOptions>,
    contentType: undefined,
    include: undefined,
  })

  const schemas = oas.getDefinition().components?.schemas

  test('generate x-enum-varnames types', async () => {
    const node = generator.buildSource('enumVarNames', schemas?.['enumVarNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate x-enumNames types', async () => {
    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})
