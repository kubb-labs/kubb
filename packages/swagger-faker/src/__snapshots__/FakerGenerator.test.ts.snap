// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FakeGenerator enums > generate x-enum-varnames types 1`] = `
[
  "

export function enumVarNames(override?: NonNullable<Partial<EnumVarNames>>): NonNullable<EnumVarNames> {
  faker.seed(1)
  return faker.helpers.arrayElement<any>(["Pending", "Received"])
}
",
]
`;

exports[`FakeGenerator enums > generate x-enumNames types 1`] = `
[
  "

export function enumNames(override?: NonNullable<Partial<EnumNames>>): NonNullable<EnumNames> {
  faker.seed(1)
  return faker.helpers.arrayElement<any>(["Pending", "Received"])
}
",
]
`;
